### ISSUE TITLE
Feature Request: API for use from other languages

### ISSUE THREAD

**Issue by KrishnaPG** [vmware/concord-bft#257]
**Is your feature request related to a problem? Please describe.**
Please make it possible to use this library from other languages. Something like REST/HTTP or Websocket messaging / RPC API to consume this library from other languages, such as Javascript.

**Describe the solution you'd like**
The **state** should be capable of being defined in and manageable from any language. For example, in-browser javascript, we define the state object that uses the underlying `concord-bft` consens engine to sync the state with other browser-based instance, native-app node instances etc.

- The app talks to underlying consensus engine over RPC (over http / websockets, for example: [BRPC](https://github.com/apache/incubator-brpc) or [DHT-RPC](https://github.com/mafintosh/dht-rpc)). 
- Multiple app instances can join the consensus network on the fly and leave (e.g. new browser tab starts another app instance with the replicated state). The DHT-RPC makes it possible to do the decentralized P2P RPC
- Apps store their state data in databases of their choice, either in local files, or distributed databases, such as MongoDB, ElasticSearch clusters on AWS
- The `concord-bft` does not fix the schema of the state object. Using serialization such as [cbor ](https://cbor.io/), instead to send and receive the state data during the RPC.

**Describe alternatives you've considered**
Tendermint ABCI is a good reference for inspiration. It has a clear separation between application state (managed by the app in any language, called ABCI proxy) and blockchain-state (managed by the Tendermint Node for consensus). Ref. [Lotion.JS](https://github.com/nomic-io/lotion)

**Additional context**
A consensus engine that can replicate multiple state-machines across multiple applications, created in  multiple languages.

**Comment by salieri11** [comment]
There is an easier option: to implement SimpleClient in any language. You need to implement the communication module as well and this is the most safe way from the BFT perspective, since you always trust your BFT client as it is within your code.

**Comment by teoparvanov** [comment]
I would just like to add that we already have a "default" Python client implementation, which can be found here:
https://github.com/vmware/concord-bft/tree/master/util/pyclient

More implementations could be added going forward.

**Comment by KrishnaPG** [comment]
Unfortunately, tight-couping through integrated APIs is not as effective for cross-language support as exposing REST/WebSocket/RPC API. 

Think of a micro-services kind of paradigm for cross-language development.

Tight-coupling is reminiscent of the monolithic design paradigm from the earlier decades. If it is, somehow, possible to expose the BFT as RPC invocable across different languages, it would greatly help with scaling, ease of use and large-scale adoption. All major languages currently support JSON-RPC or any of the HTTP equivalent RPCs.

**Comment by salieri11** [comment]
I agree about the paradigm, but when talking about BFT systems, when "owning" BFT client means having trust - its different. Think of micro service, running internally BFT client and exposing some kind of well known API outside. The caller of this service - how it can know that the reply actually came from the Concord BFT cluster? Maybe the microservice itself is malicious?

**Comment by KrishnaPG** [comment]
Thank you @salieri11 

You are right if it were a _service_. Here we are not talking about a service that is exposed outside. This is just a micro-service that is available only for the App (usually through loopback IP).

In fact, as stated earlier, it is a "microservice kind of paradigm", not micro-service exactly, since only one App would be connected to one BFT engine "micro-service".

In other words, this works exactly like the current tight-coupled language binding - only that the tight-coupling happens over RPC. Each app will have its own BFT engine instance running separately (no BFT cluster), but the App and BFT engine are not "compiled" into the same executable, they are just two processes connected over RPC (probably over loopback or IPC or unix domain sockets). The advantage being, language neutrality. Think of Unix _piping_ architecture (the grand monther of all micro-service architectures).

Tendermint ABCI is a good case-study to get more details on how this RPC mechanism works.

**Comment by KrishnaPG** [comment]
I have covered the subject in much more depth in relation to **multi-tenancy** with BFT engines here: https://github.com/tendermint/tendermint/issues/4058 

It also addresses the issue of malicious actors in the cluster, noted above by @salieri11

**Comment by salieri11** [comment]
@KrishnaPG Concord-BFT replica is absolutely language-agnostic. It doesnt assume anything about the state itself, how its managed and what language does it use. What you suggest is to use RPC or another communication layer, between the application and the BFT CLIENT. Note that application doesn't communicate to the replica directly, only via the BFT client.
The library provides upcall to the application, for the execution phase. If I understand correctly, you suggest to remove the execution from being in-proc and to allow the library to pass the execution data to some other running process, maybe remote (using RPC).
So there are 2 different things here: first, how the external client sends the request and then how the library delegates the execution to the application.
This is what you are trying to do, am I right?

**Comment by KrishnaPG** [comment]
Thank you @salieri11 

Overall we seem to be aligned, or pretty close. 

I am not able to tell the difference between _application_ and _client_ per se (Unfortunately, ReadMe is not much clear on those terms, and also in P2P the "client" is somewhat confusing role, since any and every app / machine / library can be a client, or server at will).

To help the situation I tried to summarize my understanding in the below diagram at the end (hopefully it should be openable in new tab with a larger size for clarity). Its crude, apologies for the bad artistry.

As shown in the diagram, the goal is to be able to **consume** the Concord **BFT Engine** from different languages to make them BFT-enabled. For example,

- an eCommerce application written in NodeJS, moving from centralized DB --> to BFT engine, where buyers and sellers all form the eCommerce BFT network. (A third-party could be providing value-add services on the BFT data, such as searching for active Gift-coupons, deals etc.)
- a file-sharing app written in C++ that uses BFT engine to track the current owner of a file and who all have access to it.

Now merge these two, and you get eCommerce for digital files (selling and buying MP3, Videos etc.) or pay-per-file-hosting such as drop-box etc. services. 

To summarize, the goal I am trying to address is: can one take a centralized DB app of today, and move it to Cocord BFT engine, by replacing the DB layer with BFT Engine API layer. At the end of the day, BFT is just distributed DB. Just like ODBC/JDBC allows app in any language to do CRUD/ACID transactions, BFT engine should some API that allows _any language_ to take advantage of the reading/writing/proposing blocks to the BFT. 

In other words, separate the BFT engine (consensus logic) from the business logic (application in any language).

![image](https://user-images.githubusercontent.com/4610831/69909285-caaaff80-141e-11ea-9a44-f017c8a25ceb.png)

**Comment by salieri11** [comment]
I understand. I think that its doable event today, and the only thing you need to implement is BFT client (see SimpleClient.hpp) in each of the "BFT Enabled Application" boxes, in its dedicated language. Another option will be to run on each machine BFT Client + Bft Replica (e.g. in the box called "Concord BFT engine" and to communicate with this box by any communication implementation (even POSIX sockets). The "BFT engine" box should have some CommandHandler code, written in C++ (and this is NOT the part of the BFT engine, logically, but the part of the execution), which will implement any persistent persistent storage.
The only messages that are passed to the "BFT engine" are generic requests, which payload is absolutely opaque to the BFt engine itself.
Hope now it makes sense to you. If you feel we need to fill more gaps here I suggest to schedule a Zoom call further on

**Comment by KrishnaPG** [comment]
Thank you @salieri11 

The SimpleClient.hpp seems simple enough. If all that it takes is implementing that interface, then it is a great work indeed. Just two points:

1.  Where can one find more details about the procedure of implementing "SimpleClient.hpp"? For example, the method "sendRequest", where does it send the requests to and what is expected in return.
2. The implementers (in a different language) do not need any C/C++ toolset for this development - correct? 

If developers of different language needs C/C++ toolset to use the BFT engine, then this is the wrong approach. The goal of this feature request is to make it plug & play, so that the developers from other language do not need to know C/C++ to take advantage the BFT consensus.

You have already developed the consens engine, which is a great work. Just exposing it to more wider audience will take this to new heights (currently Go, Rust, Javascript are ruling the Blockchain world).

**Comment by salieri11** [comment]
@KrishnaPG Somehow I dont see your last comment here... I just got the email but where is the comment?? Weird..

**Comment by salieri11** [comment]
Closing as irrelevant
